tests/integration/helm/upgrade/util.go

```diff
@@ -0,0 +1,188 @@
+	if err := h.DeleteChart(helmtest.EgressReleaseName, helmtest.IstioNamespace); err != nil {
+		return fmt.Errorf("failed to delete %s release", helmtest.EgressReleaseName)
+	}
+	if err := h.DeleteChart(helmtest.IngressReleaseName, helmtest.IstioNamespace); err != nil {
+		return fmt.Errorf("failed to delete %s release", helmtest.IngressReleaseName)
+	}
+	if err := h.DeleteChart(helmtest.IstiodReleaseName, helmtest.IstioNamespace); err != nil {
+		return fmt.Errorf("failed to delete %s release", helmtest.IngressReleaseName)
+	}
+	if err := h.DeleteChart(helmtest.BaseReleaseName, helmtest.IstioNamespace); err != nil {
+		return fmt.Errorf("failed to delete %s release", helmtest.BaseReleaseName)
+	}
+	if err := cs.CoreV1().Namespaces().Delete(context.TODO(), helmtest.IstioNamespace, metav1.DeleteOptions{}); err != nil {
+		return fmt.Errorf("failed to delete istio namespace: %v", err)
+	}
+	if err := kubetest.WaitForNamespaceDeletion(cs, helmtest.IstioNamespace, retry.Timeout(helmtest.RetryTimeOut)); err != nil {
+		return fmt.Errorf("wating for istio namespace to be deleted: %v", err)
+	}
+
+	return nil
+}
+
+// GetValuesOverrides returns the the values file created to pass into Helm override default values
+// for the hub and tag
+func GetValuesOverrides(ctx framework.TestContext, valuesStr, hub, tag string) string {
+	workDir := ctx.CreateTmpDirectoryOrFail("helm")
+	overrideValues := fmt.Sprintf(valuesStr, hub, tag)
+	overrideValuesFile := filepath.Join(workDir, "values.yaml")
+	if err := ioutil.WriteFile(overrideValuesFile, []byte(overrideValues), os.ModePerm); err != nil {
+		ctx.Fatalf("failed to write iop cr file: %v", err)
+	}
+
+	return overrideValuesFile
+}
+
+// InstallIstio install Istio using Helm charts with the provided
+// override values file and fails the tests on any failures.
+func InstallIstio(t *testing.T, cs cluster.Cluster,
+	h *helm.Helm, overrideValuesFile string) {
+	helmtest.CreateNamespace(t, cs, helmtest.IstioNamespace)
+
+	// Install base chart
+	err := h.InstallChart(helmtest.BaseReleaseName, helmtest.BaseChart+helmtest.TarGzSuffix,
+		helmtest.IstioNamespace, overrideValuesFile, helmtest.HelmTimeout)
+	if err != nil {
+		t.Errorf("failed to install istio %s chart", helmtest.BaseChart)
+	}
+
+	// Install discovery chart
+	err = h.InstallChart(helmtest.IstiodReleaseName, filepath.Join(helmtest.ControlChartsDir, helmtest.DiscoveryChart)+helmtest.TarGzSuffix,
+		helmtest.IstioNamespace, overrideValuesFile, helmtest.HelmTimeout)
+	if err != nil {
+		t.Errorf("failed to install istio %s chart", helmtest.DiscoveryChart)
+	}
+
+	helmtest.InstallGatewaysCharts(t, cs, h, helmtest.TarGzSuffix, helmtest.IstioNamespace, overrideValuesFile)
+}
+
+// PerformUpgradeFunc returns the provided function necessary to run inside of a integration test
+// for upgrade capability
+func PerformUpgradeFunc(t *testing.T, previousVersion string) func(framework.TestContext) {
+	return func(ctx framework.TestContext) {
```

Comment:
nit: could we replace `ctx framework.TestContext` with `t framework.TestContext` here to shadow `testing.T`? Been doing this across all our integration tests

Reply:
Or actually, ideally, `PerformUpgradeFunc` wouldn't even take `t *testing.T`

Reply:
Updated.